modules/account-lib/src/coin/hbar/transaction.ts

```diff
@@ -27,96 +35,119 @@ export class Transaction extends BaseTransaction {
+    this.innerTransaction(SDKTransaction.fromBytes(innerTx.serializeBinary()));
+
+    // Previous signatures are kept so we just add the new signature to the list
     this._signatures.push(signature);
   }
 
+  /**
+   * Replace transaction signatures with the ones from
+   * the inner SDK transaction
+   */
+  private replaceSignatures(): void {
+    this._signatures = this._hederaTx
+      ._toProto()
+      .getSigmap()!
+      .getSigpairList()
+      .map(sigPair => {
+        return toHex(sigPair.getEd25519_asU8());
+      });
+  }
+
+  /**
+   * Retrieve the signatures from the inner transaction
+   *
+   * @returns {SignaturePair[]} A list of signature pairs
+   */
+  retrieveSignatures(): SignaturePair[] {
+    return this._hederaTx && this._hederaTx._toProto().getSigmap()
+      ? this._hederaTx
+          ._toProto()!
+          .getSigmap()!
+          .getSigpairList()
+      : [];
+  }
+
+  /**
+   * Get the inner transaction body deserialized in the SDK
+   * proto default format
+   *
+   * @returns {TransactionBody} a transaction body object
+   */
+  txBody(): TransactionBody {
+    return TransactionBody.deserializeBinary(this._hederaTx._toProto().getBodybytes_asU8());
+  }
 
   /** @inheritdoc */
   toBroadcastFormat(): string {
-    return toHex(this.encode(this._hederaTx));
+    return toHex(this._hederaTx.toBytes());
   }
 
   /** @inheritdoc */
   toJson(): TxData {
-    const [acc, time] = this.getTxIdParts();
+    const txData = this._hederaTx._toProto().toObject();
+    const txBody = this.txBody().toObject();
+
+    const acc = stringifyAccountId(txBody.transactionid!.accountid!);
+    const time = stringifyTxTime(txBody.transactionid!.transactionvalidstart!);
     const result: TxData = {
       id: acc + '@' + time,
-      hash: this.getTxHash(), // TODO: Update once hedera-sdk release this functionality BGA-284
-      data: toHex(this._hederaTx.bodyBytes),
-      fee: new BigNumber(this._txBody.transactionFee!.toString()).toNumber(),
+      hash: txData.sigmap ? toHex(this._hederaTx.hash()) : '', // Hash is returned if the transaction was signed
+      data: toHex(this._hederaTx._toProto().getBodybytes_asU8()),
+      fee: new BigNumber(txBody.transactionfee).toNumber(),
       from: acc,
       startTime: time,
-      validDuration: this._txBody.transactionValidDuration!.seconds!.toString(),
-      node: stringifyAccountId(this._txBody.nodeAccountID!),
-      memo: this._txBody.memo,
+      validDuration: txBody.transactionvalidduration!.seconds.toString(),
+      node: stringifyAccountId(txBody.nodeaccountid!),
+      memo: txBody.memo,
     };
 
-    if (this._txBody.data === 'cryptoTransfer') {
-      const [recipient, amount] = this.getTransferData();
-      result.amount = amount;
-      result.to = recipient;
+    if (txBody.cryptotransfer) {
+      txBody.cryptotransfer!.transfers!.accountamountsList.forEach(transfer => {
+        if (Long.fromValue(transfer.amount).isPositive()) {
```

Comment:
can these be negative? or 0? we should probably throw if they are

Reply:
If it's negative that would be the transfer recipient. They could be zero but we haven't tried that. In that case we wouldn't be able to tell which is the sender and which is the recipient.

Reply:
If there is no positive amount then the transaction is probably wrong because every transaction should have the same amount of coin input and output (i.e. if there is a 5 negative, there should be a 5 positive)

Reply:
After further research we've found that the best alternative for this would be to look into the transfers for an account different from the source account (the account included in the tx ID) and set that as the transfer recipient.